// @thisissoon/angular-scrollspy/esm2015/app/scroll-spy/scroll-spy-service/scroll-spy.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
/**
 * Service that stores a list of `Spy`'s and the state
 * of their nav items `inViewport` and `active` state
 *
 */
export class ScrollSpyService {
    constructor() {
        /**
         * List of `Spy`'s
         *
         * \@memberof ScrollSpyService
         */
        this.spys = [];
        /**
         * Stores requests to add items to spy when spy hasn't been created
         * yet. Once spy has been added then request will be made again.
         *
         * \@memberof ScrollSpyService
         */
        this.buffer = [];
    }
    /**
     * Add spy to list of `spys`
     *
     * \@memberof ScrollSpyService
     * @param {?} id
     * @param {?} items
     * @return {?}
     */
    addSpy(id, items) {
        this.spys.push({ id, items });
        /** @type {?} */
        const buffer = this.buffer.filter(i => i.spyId === id);
        this.buffer = this.buffer.filter(i => i.spyId !== id);
        buffer.forEach(i => this.setSpySectionStatus(i.sectionId, i.spyId, i.inViewport));
    }
    /**
     * Remove spy from list of `spys`
     *
     * \@memberof ScrollSpyService
     * @param {?} id
     * @return {?}
     */
    removeSpy(id) {
        /** @type {?} */
        const i = this.spys.findIndex(s => s.id === id);
        this.spys.splice(i, 1);
    }
    /**
     * Set the `inViewport` status for a spy item then sets the active
     * to true for the first item in the list that has `inViewport`
     * set to true
     *
     * \@memberof ScrollSpyService
     * @param {?} sectionId
     * @param {?} spyId
     * @param {?} inViewport
     * @return {?}
     */
    setSpySectionStatus(sectionId, spyId, inViewport) {
        /** @type {?} */
        const spy = this.spys.find(s => s.id === spyId);
        if (!spy) {
            this.buffer.push({ sectionId, spyId, inViewport });
            return;
        }
        /** @type {?} */
        const item = spy.items.find(i => i.section === sectionId);
        if (!item) {
            return;
        }
        item.inViewport = inViewport;
        /** @type {?} */
        const firstInViewport = spy.items.filter(i => i.inViewport)[0];
        spy.items.forEach(i => (i.active = false));
        if (firstInViewport) {
            firstInViewport.active = true;
            firstInViewport.detectChanges();
        }
    }
}
ScrollSpyService.decorators = [
    { type: Injectable }
];
if (false) {
    /**
     * List of `Spy`'s
     *
     * \@memberof ScrollSpyService
     * @type {?}
     */
    ScrollSpyService.prototype.spys;
    /**
     * Stores requests to add items to spy when spy hasn't been created
     * yet. Once spy has been added then request will be made again.
     *
     * \@memberof ScrollSpyService
     * @type {?}
     */
    ScrollSpyService.prototype.buffer;
}
