// @thisissoon/angular-scrollspy/esm5/app/scroll-spy/scroll-spy-item/scroll-spy-item.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, HostBinding, ChangeDetectorRef, } from '@angular/core';
/**
 * A directive used to add an `active` class to a nav item
 * when the section is in the viewport
 *
 * \@example
 * ```html
 * <a snScrollSpyItem for="foo" href="#section1">Section 1</a>
 * ```
 *
 */
var ScrollSpyItemDirective = /** @class */ (function () {
    /**
     * Creates an instance of ScrollSpyItemDirective.
     * @memberof ScrollSpyItemDirective
     */
    function ScrollSpyItemDirective(cdRef) {
        this.cdRef = cdRef;
        /**
         * True if the nav item is the active item in the `items` list
         * for `ScrollSpyDirective` instance
         *
         * \@memberof ScrollSpyItemDirective
         */
        this.active = false;
        /**
         * If true means the section is in the viewport
         *
         * \@memberof ScrollSpyItemDirective
         */
        this.inViewport = false;
    }
    Object.defineProperty(ScrollSpyItemDirective.prototype, "section", {
        /**
         * Id of section that links navigates to
         *
         * @readonly
         * @memberof ScrollSpyItemDirective
         */
        get: /**
         * Id of section that links navigates to
         *
         * \@readonly
         * \@memberof ScrollSpyItemDirective
         * @return {?}
         */
        function () {
            return this.href.replace('#', '');
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Manually trigger change detection
     *
     * @memberof ScrollSpyItemDirective
     */
    /**
     * Manually trigger change detection
     *
     * \@memberof ScrollSpyItemDirective
     * @return {?}
     */
    ScrollSpyItemDirective.prototype.detectChanges = /**
     * Manually trigger change detection
     *
     * \@memberof ScrollSpyItemDirective
     * @return {?}
     */
    function () {
        this.cdRef.detectChanges();
    };
    ScrollSpyItemDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[snScrollSpyItem]',
                },] }
    ];
    /** @nocollapse */
    ScrollSpyItemDirective.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    ScrollSpyItemDirective.propDecorators = {
        active: [{ type: HostBinding, args: ['class.active',] }],
        for: [{ type: Input }],
        href: [{ type: Input }]
    };
    return ScrollSpyItemDirective;
}());
export { ScrollSpyItemDirective };
if (false) {
    /**
     * True if the nav item is the active item in the `items` list
     * for `ScrollSpyDirective` instance
     *
     * \@memberof ScrollSpyItemDirective
     * @type {?}
     */
    ScrollSpyItemDirective.prototype.active;
    /**
     * ID of `ScrollSpyDirective` instance
     *
     * \@memberof ScrollSpyItemDirective
     * @type {?}
     */
    ScrollSpyItemDirective.prototype.for;
    /**
     * Hash for section to link to
     *
     * \@memberof ScrollSpyItemDirective
     * @type {?}
     */
    ScrollSpyItemDirective.prototype.href;
    /**
     * If true means the section is in the viewport
     *
     * \@memberof ScrollSpyItemDirective
     * @type {?}
     */
    ScrollSpyItemDirective.prototype.inViewport;
    /**
     * @type {?}
     * @private
     */
    ScrollSpyItemDirective.prototype.cdRef;
}
