// @thisissoon/angular-scrollspy/esm5/app/scroll-spy/scroll-spy/scroll-spy.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ContentChildren, QueryList, Input, } from '@angular/core';
import { ScrollSpyItemDirective } from '../scroll-spy-item/scroll-spy-item.directive';
import { ScrollSpyService } from '../scroll-spy-service/scroll-spy.service';
/**
 * Adds `active` class to navigation links when section is in the viewport.
 * Used in conjuction with `snScrollItem` directive which should be added
 * to anchor links in the nav
 *
 * \@example
 * ```
 * <ul role="navigation" snScrollSpy id="foo">
 *  <li><a snScrollSpyItem for="foo" href="#section1">Section 1</a></li>
 *  <li><a snScrollSpyItem for="foo" href="#section2">Section 2</a></li>
 *  <li><a snScrollSpyItem for="foo" href="#section3">Section 3</a></li>
 *  <li><a snScrollSpyItem for="foo" href="#section4">Section 4</a></li>
 * </ul>
 * ```
 *
 */
var ScrollSpyDirective = /** @class */ (function () {
    /**
     * Creates an instance of ScrollSpyDirective.
     * @memberof ScrollSpyDirective
     */
    function ScrollSpyDirective(scrollSpySvc) {
        this.scrollSpySvc = scrollSpySvc;
    }
    /**
     * Adds spy to list of spys in `ScrollSpyService`
     *
     * @memberof ScrollSpyDirective
     */
    /**
     * Adds spy to list of spys in `ScrollSpyService`
     *
     * \@memberof ScrollSpyDirective
     * @return {?}
     */
    ScrollSpyDirective.prototype.ngAfterViewInit = /**
     * Adds spy to list of spys in `ScrollSpyService`
     *
     * \@memberof ScrollSpyDirective
     * @return {?}
     */
    function () {
        this.scrollSpySvc.addSpy(this.id, this.items);
    };
    /**
     * Remove spy from list of spys when directive is destroyed
     *
     * @memberof ScrollSpyDirective
     */
    /**
     * Remove spy from list of spys when directive is destroyed
     *
     * \@memberof ScrollSpyDirective
     * @return {?}
     */
    ScrollSpyDirective.prototype.ngOnDestroy = /**
     * Remove spy from list of spys when directive is destroyed
     *
     * \@memberof ScrollSpyDirective
     * @return {?}
     */
    function () {
        this.scrollSpySvc.removeSpy(this.id);
    };
    ScrollSpyDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[snScrollSpy]',
                },] }
    ];
    /** @nocollapse */
    ScrollSpyDirective.ctorParameters = function () { return [
        { type: ScrollSpyService }
    ]; };
    ScrollSpyDirective.propDecorators = {
        items: [{ type: ContentChildren, args: [ScrollSpyItemDirective,] }],
        id: [{ type: Input }]
    };
    return ScrollSpyDirective;
}());
export { ScrollSpyDirective };
if (false) {
    /**
     * Collection of `ScrollSpyItem`. They are the list of
     * nav items.
     *
     * \@memberof ScrollSpyDirective
     * @type {?}
     */
    ScrollSpyDirective.prototype.items;
    /**
     * ID of scrollSpy instance
     *
     * \@memberof ScrollSpyDirective
     * @type {?}
     */
    ScrollSpyDirective.prototype.id;
    /**
     * @type {?}
     * @private
     */
    ScrollSpyDirective.prototype.scrollSpySvc;
}
