// @thisissoon/angular-scrollspy/esm5/app/scroll-spy/scroll-spy.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { InViewportModule } from '@thisissoon/angular-inviewport';
import { ScrollSpyDirective } from './scroll-spy/scroll-spy.directive';
import { ScrollSpySectionComponent } from './scroll-spy-section/scroll-spy-section.component';
import { ScrollSpyItemDirective } from './scroll-spy-item/scroll-spy-item.directive';
import { ScrollSpyService } from './scroll-spy-service/scroll-spy.service';
/** @type {?} */
var directives = [ScrollSpyDirective, ScrollSpyItemDirective];
/** @type {?} */
var components = [ScrollSpySectionComponent];
/** @type {?} */
var providers = [ScrollSpyService];
/**
 * A simple lightweight library for Angular which automatically
 * updates links to indicate the currently active section in the viewport
 *
 */
var ScrollSpyModule = /** @class */ (function () {
    function ScrollSpyModule() {
    }
    /**
     * Specify a static method for root module to ensure providers are
     * only provided once but allows the module to still be imported
     * into other modules without reproviding services.
     *
     * @memberof ScrollSpyModule
     */
    /**
     * Specify a static method for root module to ensure providers are
     * only provided once but allows the module to still be imported
     * into other modules without reproviding services.
     *
     * \@memberof ScrollSpyModule
     * @return {?}
     */
    ScrollSpyModule.forRoot = /**
     * Specify a static method for root module to ensure providers are
     * only provided once but allows the module to still be imported
     * into other modules without reproviding services.
     *
     * \@memberof ScrollSpyModule
     * @return {?}
     */
    function () {
        return {
            ngModule: ScrollSpyModule,
            providers: tslib_1.__spread(providers),
        };
    };
    ScrollSpyModule.decorators = [
        { type: NgModule, args: [{
                    imports: [InViewportModule],
                    declarations: tslib_1.__spread(directives, components),
                    exports: tslib_1.__spread(directives, components),
                },] }
    ];
    return ScrollSpyModule;
}());
export { ScrollSpyModule };
