// @thisissoon/angular-scrollspy/esm5/app/scroll-spy/scroll-spy-service/scroll-spy.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
/**
 * Service that stores a list of `Spy`'s and the state
 * of their nav items `inViewport` and `active` state
 *
 */
var ScrollSpyService = /** @class */ (function () {
    function ScrollSpyService() {
        /**
         * List of `Spy`'s
         *
         * \@memberof ScrollSpyService
         */
        this.spys = [];
        /**
         * Stores requests to add items to spy when spy hasn't been created
         * yet. Once spy has been added then request will be made again.
         *
         * \@memberof ScrollSpyService
         */
        this.buffer = [];
    }
    /**
     * Add spy to list of `spys`
     *
     * @memberof ScrollSpyService
     */
    /**
     * Add spy to list of `spys`
     *
     * \@memberof ScrollSpyService
     * @param {?} id
     * @param {?} items
     * @return {?}
     */
    ScrollSpyService.prototype.addSpy = /**
     * Add spy to list of `spys`
     *
     * \@memberof ScrollSpyService
     * @param {?} id
     * @param {?} items
     * @return {?}
     */
    function (id, items) {
        var _this = this;
        this.spys.push({ id: id, items: items });
        /** @type {?} */
        var buffer = this.buffer.filter(function (i) { return i.spyId === id; });
        this.buffer = this.buffer.filter(function (i) { return i.spyId !== id; });
        buffer.forEach(function (i) {
            return _this.setSpySectionStatus(i.sectionId, i.spyId, i.inViewport);
        });
    };
    /**
     * Remove spy from list of `spys`
     *
     * @memberof ScrollSpyService
     */
    /**
     * Remove spy from list of `spys`
     *
     * \@memberof ScrollSpyService
     * @param {?} id
     * @return {?}
     */
    ScrollSpyService.prototype.removeSpy = /**
     * Remove spy from list of `spys`
     *
     * \@memberof ScrollSpyService
     * @param {?} id
     * @return {?}
     */
    function (id) {
        /** @type {?} */
        var i = this.spys.findIndex(function (s) { return s.id === id; });
        this.spys.splice(i, 1);
    };
    /**
     * Set the `inViewport` status for a spy item then sets the active
     * to true for the first item in the list that has `inViewport`
     * set to true
     *
     * @memberof ScrollSpyService
     */
    /**
     * Set the `inViewport` status for a spy item then sets the active
     * to true for the first item in the list that has `inViewport`
     * set to true
     *
     * \@memberof ScrollSpyService
     * @param {?} sectionId
     * @param {?} spyId
     * @param {?} inViewport
     * @return {?}
     */
    ScrollSpyService.prototype.setSpySectionStatus = /**
     * Set the `inViewport` status for a spy item then sets the active
     * to true for the first item in the list that has `inViewport`
     * set to true
     *
     * \@memberof ScrollSpyService
     * @param {?} sectionId
     * @param {?} spyId
     * @param {?} inViewport
     * @return {?}
     */
    function (sectionId, spyId, inViewport) {
        /** @type {?} */
        var spy = this.spys.find(function (s) { return s.id === spyId; });
        if (!spy) {
            this.buffer.push({ sectionId: sectionId, spyId: spyId, inViewport: inViewport });
            return;
        }
        /** @type {?} */
        var item = spy.items.find(function (i) { return i.section === sectionId; });
        if (!item) {
            return;
        }
        item.inViewport = inViewport;
        /** @type {?} */
        var firstInViewport = spy.items.filter(function (i) { return i.inViewport; })[0];
        spy.items.forEach(function (i) { return (i.active = false); });
        if (firstInViewport) {
            firstInViewport.active = true;
            firstInViewport.detectChanges();
        }
    };
    ScrollSpyService.decorators = [
        { type: Injectable }
    ];
    return ScrollSpyService;
}());
export { ScrollSpyService };
if (false) {
    /**
     * List of `Spy`'s
     *
     * \@memberof ScrollSpyService
     * @type {?}
     */
    ScrollSpyService.prototype.spys;
    /**
     * Stores requests to add items to spy when spy hasn't been created
     * yet. Once spy has been added then request will be made again.
     *
     * \@memberof ScrollSpyService
     * @type {?}
     */
    ScrollSpyService.prototype.buffer;
}
