// @thisissoon/angular-inviewport/esm5/app/in-viewport/in-viewport.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { InViewportDirective } from './in-viewport.directive';
import { WINDOW } from '../window/window-token';
import { WINDOW_MOCK } from '../window/window-mock';
/**
 * A simple lightweight library for Angular with other dependencies
 * that detects when an element is within the browsers viewport and adds a
 * sn-viewport--in or sn-viewport--out class to the element.
 */
var InViewportModule = /** @class */ (function () {
    function InViewportModule() {
    }
    /**
     * @return {?}
     */
    InViewportModule.forServer = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: InViewportModule,
            providers: [{ provide: WINDOW, useValue: WINDOW_MOCK }]
        };
    };
    InViewportModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [InViewportDirective],
                    exports: [InViewportDirective]
                },] }
    ];
    return InViewportModule;
}());
export { InViewportModule };
