// @thisissoon/angular-scrollspy/esm5/app/scroll-spy/scroll-spy-section/scroll-spy-section.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { ScrollSpyService } from '../scroll-spy-service/scroll-spy.service';
/**
 * A component to wrap section content within that will update the
 * `ScrollSpyService` when it's in view
 *
 * \@example
 * ```html
 *  <sn-scroll-spy-section id="section1" for="foo">
 *    ...
 *  </sn-scroll-spy-section>
 * ```
 */
var ScrollSpySectionComponent = /** @class */ (function () {
    /**
     * Creates an instance of ScrollSpySectionComponent.
     * @memberof ScrollSpySectionComponent
     */
    function ScrollSpySectionComponent(scrollSpySvc) {
        this.scrollSpySvc = scrollSpySvc;
    }
    /**
     * Updates `ScrollSpy` section when element enters/leaves viewport
     *
     * @memberof ScrollSpySectionComponent
     */
    /**
     * Updates `ScrollSpy` section when element enters/leaves viewport
     *
     * \@memberof ScrollSpySectionComponent
     * @param {?} inViewport
     * @return {?}
     */
    ScrollSpySectionComponent.prototype.onInViewportChange = /**
     * Updates `ScrollSpy` section when element enters/leaves viewport
     *
     * \@memberof ScrollSpySectionComponent
     * @param {?} inViewport
     * @return {?}
     */
    function (inViewport) {
        this.scrollSpySvc.setSpySectionStatus(this.id, this.for, inViewport);
    };
    ScrollSpySectionComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sn-scroll-spy-section',
                    template: "<div\n  class=\"sn-hidden\"\n  snInViewport\n  (inViewportChange)=\"onInViewportChange($event)\">\n</div>\n<ng-content></ng-content>\n",
                    styles: [":host{display:block;position:relative}.sn-hidden{bottom:0;left:0;opacity:0;position:absolute;right:0;top:0;z-index:-1}"]
                }] }
    ];
    /** @nocollapse */
    ScrollSpySectionComponent.ctorParameters = function () { return [
        { type: ScrollSpyService }
    ]; };
    ScrollSpySectionComponent.propDecorators = {
        id: [{ type: Input }],
        for: [{ type: Input }]
    };
    return ScrollSpySectionComponent;
}());
export { ScrollSpySectionComponent };
if (false) {
    /**
     * Identifies the section
     *
     * \@memberof ScrollSpySectionComponent
     * @type {?}
     */
    ScrollSpySectionComponent.prototype.id;
    /**
     * Specifies which `ScrollSpy` instance to update
     *
     * \@memberof ScrollSpySectionComponent
     * @type {?}
     */
    ScrollSpySectionComponent.prototype.for;
    /**
     * @type {?}
     * @private
     */
    ScrollSpySectionComponent.prototype.scrollSpySvc;
}
