// @thisissoon/angular-scrollspy/esm2015/app/scroll-spy/scroll-spy-item/scroll-spy-item.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, HostBinding, ChangeDetectorRef, } from '@angular/core';
/**
 * A directive used to add an `active` class to a nav item
 * when the section is in the viewport
 *
 * \@example
 * ```html
 * <a snScrollSpyItem for="foo" href="#section1">Section 1</a>
 * ```
 *
 */
export class ScrollSpyItemDirective {
    /**
     * Creates an instance of ScrollSpyItemDirective.
     * \@memberof ScrollSpyItemDirective
     * @param {?} cdRef
     */
    constructor(cdRef) {
        this.cdRef = cdRef;
        /**
         * True if the nav item is the active item in the `items` list
         * for `ScrollSpyDirective` instance
         *
         * \@memberof ScrollSpyItemDirective
         */
        this.active = false;
        /**
         * If true means the section is in the viewport
         *
         * \@memberof ScrollSpyItemDirective
         */
        this.inViewport = false;
    }
    /**
     * Id of section that links navigates to
     *
     * \@readonly
     * \@memberof ScrollSpyItemDirective
     * @return {?}
     */
    get section() {
        return this.href.replace('#', '');
    }
    /**
     * Manually trigger change detection
     *
     * \@memberof ScrollSpyItemDirective
     * @return {?}
     */
    detectChanges() {
        this.cdRef.detectChanges();
    }
}
ScrollSpyItemDirective.decorators = [
    { type: Directive, args: [{
                selector: '[snScrollSpyItem]',
            },] }
];
/** @nocollapse */
ScrollSpyItemDirective.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
ScrollSpyItemDirective.propDecorators = {
    active: [{ type: HostBinding, args: ['class.active',] }],
    for: [{ type: Input }],
    href: [{ type: Input }]
};
if (false) {
    /**
     * True if the nav item is the active item in the `items` list
     * for `ScrollSpyDirective` instance
     *
     * \@memberof ScrollSpyItemDirective
     * @type {?}
     */
    ScrollSpyItemDirective.prototype.active;
    /**
     * ID of `ScrollSpyDirective` instance
     *
     * \@memberof ScrollSpyItemDirective
     * @type {?}
     */
    ScrollSpyItemDirective.prototype.for;
    /**
     * Hash for section to link to
     *
     * \@memberof ScrollSpyItemDirective
     * @type {?}
     */
    ScrollSpyItemDirective.prototype.href;
    /**
     * If true means the section is in the viewport
     *
     * \@memberof ScrollSpyItemDirective
     * @type {?}
     */
    ScrollSpyItemDirective.prototype.inViewport;
    /**
     * @type {?}
     * @private
     */
    ScrollSpyItemDirective.prototype.cdRef;
}
